doc(development-guide): add details for history mode (#84)
fivetran/fivetran_sdk: how-to-handle-history-mode-batch-files.md

### How to track historical changes?  

The following types of files are a part of the `WriteHistoryBatchRequest` gRPC call. These files must be processed in the exact order as described in the following subsections.  

#### `earliest_start_files`

In `WriteHistoryBatchRequest`, we pass a new field, `earliest_start_files`. This file contains a list of records, where for each primary key in the incoming batch, only the record with the earliest `_fivetran_start` value is included, even if multiple records with the same primary key exist in the batch.

For this file, the following operations must be implemented in the exact order as they are listed:
1. Removing of any overlapping records where existing `_fivetran_start` is greater than the `earliest_fivetran_start` timestamp value in the `earliest_start_files` file:
  
   ```sql
   DELETE FROM <schema.table> WHERE pk1 = <val> {AND  pk2 = <val>.....} AND _fivetran_start >= val(_earliest_fivetran_start);
   ```

3. Updating of the values of the history mode-specific system columns `fivetran_active` and `fivetran_end` in the destination. 
    
  ```sql

>   UPDATE <schema.table> SET fivetran_active = FALSE, _fivetran_end = t1 - 1 WHERE _fivetran_active = TRUE AND pk1 = <val> {AND  pk2 = <val>.....}`

Reviewer: I believe T1 denotes a sync time of the incoming record that is the next one in relation. to the current active records in the destination.However, I am not sure, how with multiple records with the same PK present in the batch, does the process recognizes this "historically next record". 
My logic is that all incoming records with the same PK should be compared by their _fivetran_start and the record with the earliest _fivetran_start, but still greater than _fivetran_start of the active record is this "next record" with 'T1'.
@fivetran-rishabhghosh ^^

>   UPDATE <schema.table> SET fivetran_active = FALSE, _fivetran_end = t1 - 1 WHERE _fivetran_active = TRUE AND pk1 = <val> {AND  pk2 = <val>.....}`

Reviewer 2: t1 is the fivetran_start value of the record in earliest batch file

# What is History Mode

History mode allows to capture every available version of each record from Fivetran source connectors.
In order to keep all versions of the records, three new system columns are added to tables with history mode enabled.

Column | Type | Description
--- | --- | ---
_fivetran_active | Boolean | TRUE if it is the currently active record. FALSE if it is a historical version of the record. Only one version of the record can be TRUE.
_fivetran_start | TimeStamp | The time when the record was first created or modified in the source.
_fivetran_end | TimeStamp | The value for this column depends on whether the record is active. If the record is not active, then `_fivetran_end` value will be `_fivetran_start` of the next version of the record minus 1 millisecond. If the record is deleted, then the value will be the same as the timestamp of delete operation. If the record is active, then `_fivetran_end` is set to maximum TIMESTAMP value.

### How to Maintain History?  

Following types of files are part of the `WriteHistoryBatchRequest` grpc call. These files need to be processed as described below and in the same order.  

#### `earliest_start_files`

> - In `WriteHistoryBatchRequest`, we pass a new field, `earliest_start_files`. This file will consist of a single record for each primary key and the earliest value of the _fivetran_start value in the incoming batch.

Reviewer: pass a new field, `earliest_start_files` - field or file? 
This file will consist of a single record for each primary key and the earliest value of the _fivetran_start value in the incoming batch. - what is the source of the information for each record in the file?

Author: ```suggestion
- In `WriteHistoryBatchRequest`, we pass a new parameter, `earliest_start_files`. This parameter is the list of files which consists of a single record for each primary key and the earliest value of the _fivetran_start value in the incoming batch. 
```

>   For this new file, perform following operations in order:

Reviewer: For this new file, perform following operations in order:
 - is this imperative for the user?

Author: Yes, the sequence is important else the update won't be correct. 
- Delete overlapping records 
- Update
- Upsert 
- Delete

    - **Delete overlapping records** : 
        

>       Delete query to remove any overlapping records where existing `_fivetran_start` is greater than the earliest_fivetran_start timestamp value in the `earliest_start_files` file.

Reviewer: 1. What is the <schema.table> table the DELETE query is run against - source schema? 
2. _earliest_fivetran_start - where is the value obtained from?

  
        `DELETE FROM <schema.table> WHERE pk1 = <val> {AND  pk2 = <val>.....} AND _fivetran_start >= val(_earliest_fivetran_start);`

      - Update history mode columns: Updates `fivetran_active` and `fivetran_end` value in the destination. 
    
        `UPDATE <schema.table> SET fivetran_active = FALSE, _fivetran_end = t1 - 1 WHERE _fivetran_active = TRUE AND pk1 = <val> {AND  pk2 = <val>.....}`

#### `update_files`

> - This file contains incomplete records where modified columns have actual values whereas unmodified columns have the special value `unmodified_string`. Before inserting these records in the table these records need to be completed. To complete the incoming records, wherever the value of the column is `unmodified_string` get the value from the last active row from the destination and fill that value in the unmodified column.

Reviewer: For this file, we don't state it's a part of `WriteHistoryBatchRequest` - is it not, or why we state this for `earliest_start_files`?

Reviewer: "where modified columns have actual values whereas unmodified columns have the special value `unmodified_string`. " - what is "actual values" and "unmodified columns"? what is `unmodified)_string`?

Author: > For this file, we don't state it's a part of WriteHistoryBatchRequest - is it not, or why we state this for earliest_start_files?

All these files are part of the `WriteHistoryBatchRequest`, maybe we can reframe the sentence or add a statement before describing what needs to be done with each of the batch files.

Author: > what is "actual values" and "unmodified columns"? what is unmodified)_string?

For an update query, wherever the incoming value is present that is being called "actual value", if a column is not getting updated we add an ad-hoc string `unmodified_string` to help identify if the column was updated when `update` operation is performed.

Suppose the existing table in destination is as follows:

> Id(PK) | COL1 | COL2 | _fivetran_start(PK) | _fivetran_end | _fivetran_active | _fivetran_synced

Reviewer: What are COL1 and COL2?

Author: This is an example, we can call the columns anything.

In order to keep all versions of the records, the following new system columns are added to tables with history mode enabled:

Column | Type | Description
--- | --- | ---
_fivetran_active | Boolean | TRUE if it is the currently active record. FALSE if it is a historical version of the record. Only one version of the record can be TRUE.
_fivetran_start | TimeStamp | The time when the record was first created or modified in the source.
_fivetran_end | TimeStamp | The value for this column depends on whether the record is active. If the record is not active, then `_fivetran_end` value will be `_fivetran_start` of the next version of the record minus 1 millisecond. If the record is deleted, then the value will be the same as the timestamp of the delete operation. If the record is active, then `_fivetran_end` is set to maximum TIMESTAMP value.

### How to track historical changes?  

The following types of files are a part of the `WriteHistoryBatchRequest` gRPC call. These files must be processed in the exact order as described in the following subsections.  

#### `earliest_start_files`

> In `WriteHistoryBatchRequest`, we pass a new field, `earliest_start_files`. This file contains a list of records, where for each primary key in the incoming batch, only the record with the earliest `_fivetran_start` value is included, even if multiple records with the same primary key exist in the batch.

Reviewer 2: `This file contains a list of records, where for each primary key in the incoming batch, only the record with the earliest `_fivetran_start` value is included, even if multiple records with the same primary key exist in the batch.`

technically correct but I find it a bit confusing
Can we write something like
`This file contains a single record for each primary key in the incoming batch, with the earliest `_fivetran_start`.

Reviewer: ```suggestion
In `WriteHistoryBatchRequest`, we pass a new field, `earliest_start_files`. This file contains a single record for each primary key in the incoming batch, with the earliest `_fivetran_start`.
```

For this file, the following operations must be implemented in the exact order as they are listed:

> 1. Removing of any overlapping records where existing `_fivetran_start` is greater than the `earliest_fivetran_start` timestamp value in the `earliest_start_files` file:

Reviewer 2: Removal?

Reviewer: Both are OK.
```suggestion
1. Removing any overlapping records where existing `_fivetran_start` is greater than the `earliest_fivetran_start` timestamp value in the `earliest_start_files` file:
```
